File: frontend-admin/src/services/auditLogService.ts
import axiosClient from '../api/axiosClient';
import type {
  AuditActorType,
  AuditCategory,
  AuditLog,
  AuditStatus,
} from '../types/models';

const STORAGE_KEY = 'admin_audit_logs';
const ACTOR_KEY = 'admin_audit_actor';
const MAX_ENTRIES = 2000;

/* ------------------------------------------------------------------ */
/* Actor (người thực hiện) hiện tại                                    */
/* ------------------------------------------------------------------ */

interface CurrentActor {
  id?: string | number;
  name: string;
}

export function setAuditActor(actor: { id?: string | number; name?: string } | null): void {
  try {
    if (!actor) {
      localStorage.removeItem(ACTOR_KEY);
      return;
    }
    localStorage.setItem(
      ACTOR_KEY,
      JSON.stringify({ id: actor.id, name: actor.name || 'Admin' })
    );
  } catch {
    /* localStorage không khả dụng */
  }
}

function getCurrentActor(): CurrentActor {
  try {
    const raw = localStorage.getItem(ACTOR_KEY);
    if (raw) return JSON.parse(raw) as CurrentActor;
  } catch {
    /* ignore */
  }
  return { name: 'Admin' };
}

/* ------------------------------------------------------------------ */
/* Lưu trữ cục bộ                                                      */
/* ------------------------------------------------------------------ */

function readAll(): AuditLog[] {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return [];
    const parsed = JSON.parse(raw);
    return Array.isArray(parsed) ? (parsed as AuditLog[]) : [];
  } catch {
    return [];
  }
}

function writeAll(logs: AuditLog[]): void {
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(logs.slice(0, MAX_ENTRIES)));
  } catch {
    /* vượt quota -> bỏ qua */
  }
}

function makeId(): string {
  return `${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
}

/** Ghi một bản ghi nhật ký (mới nhất lên đầu). */
export function addLog(
  entry: Omit<AuditLog, 'id' | 'timestamp'> & { timestamp?: string }
): AuditLog {
  const log: AuditLog = {
    id: makeId(),
    timestamp: entry.timestamp || new Date().toISOString(),
    ...entry,
  };
  const all = readAll();
  all.unshift(log);
  writeAll(all);
  return log;
}

/* ------------------------------------------------------------------ */
/* Ánh xạ request -> hành động có ý nghĩa                              */
/* ------------------------------------------------------------------ */

interface ActionDescriptor {
  action: string;
  category: AuditCategory;
  /** label gốc, sẽ được bổ sung tên đối tượng khi build mô tả */
  label: string;
  targetType?: string;
  /** lấy id đối tượng từ path/body */
  targetId?: (path: string, body: any) => string | number | undefined;
}

const num = (v: string | undefined) => (v == null ? undefined : v);

/** Danh sách matcher theo method + regex path. */
const MATCHERS: Array<{
  method: string;
  re: RegExp;
  build: (m: RegExpMatchArray, body: any) => ActionDescriptor;
}> = [
  // --- AUTH ---
  {
    method: 'POST',
    re: /\/auth\/login$/,
    build: () => ({ action: 'LOGIN', category: 'AUTH', label: 'Đăng nhập quản trị' }),
  },
  {
    method: 'POST',
    re: /\/auth\/logout$/,
    build: () => ({ action: 'LOGOUT', category: 'AUTH', label: 'Đăng xuất khỏi hệ thống' }),
  },
  // --- USER ---
  {
    method: 'POST',
    re: /\/admin\/lock\/(\w+)/,
    build: (m) => ({
      action: 'LOCK_USER',
      category: 'USER',
      label: 'Khoá tài khoản người dùng',
      targetType: 'USER',
      targetId: () => num(m[1]),
    }),
  },
  {
    method: 'POST',
    re: /\/admin\/unlock\/(\w+)/,
    build: (m) => ({
      action: 'UNLOCK_USER',
      category: 'USER',
      label: 'Mở khoá tài khoản người dùng',
      targetType: 'USER',
      targetId: () => num(m[1]),
    }),
  },
  {
    method: 'DELETE',
    re: /\/auth\/users\/(\w+)/,
    build: (m) => ({
      action: 'DELETE_USER',
      category: 'USER',
      label: 'Xoá tài khoản người dùng',
      targetType: 'USER',
      targetId: () => num(m[1]),
    }),
  },
  {
    method: 'PUT',
    re: /\/auth\/users\/(\w+)/,
    build: (m) => ({
      action: 'UPDATE_USER',
      category: 'USER',
      label: 'Cập nhật thông tin người dùng',
      targetType: 'USER',
      targetId: () => num(m[1]),
    }),
  },
  // --- PAGE ---
  {
    method: 'DELETE',
    re: /\/page\/delete\/(\w+)/,
    build: (m) => ({
      action: 'DELETE_PAGE',
      category: 'PAGE',
      label: 'Xoá trang (Page)',
      targetType: 'PAGE',
      targetId: () => num(m[1]),
    }),
  },
  {
    method: 'POST',
    re: /\/page-member\/approve-join$/,
    build: (_m, b) => ({
      action: 'APPROVE_MEMBER',
      category: 'PAGE',
      label: 'Duyệt yêu cầu tham gia trang',
      targetType: 'PAGE',
      targetId: () => b?.pageId,
    }),
  },
  {
    method: 'POST',
    re: /\/page-member\/reject-join$/,
    build: (_m, b) => ({
      action: 'REJECT_MEMBER',
      category: 'PAGE',
      label: 'Từ chối yêu cầu tham gia trang',
      targetType: 'PAGE',
      targetId: () => b?.pageId,
    }),
  },
  {
    method: 'POST',
    re: /\/page-member\/block$/,
    build: (_m, b) => ({
      action: 'BLOCK_MEMBER',
      category: 'PAGE',
      label: 'Chặn thành viên trang',
      targetType: 'PAGE',
      targetId: () => b?.pageId,
    }),
  },
  {
    method: 'POST',
    re: /\/page-member\/cancel-block$/,
    build: (_m, b) => ({
      action: 'UNBLOCK_MEMBER',
      category: 'PAGE',
      label: 'Bỏ chặn thành viên trang',
      targetType: 'PAGE',
      targetId: () => b?.pageId,
    }),
  },
  {
    method: 'POST',
    re: /\/page-member\/delete$/,
    build: (_m, b) => ({
      action: 'REMOVE_MEMBER',
      category: 'PAGE',
      label: 'Xoá thành viên khỏi trang',
      targetType: 'PAGE',
      targetId: () => b?.pageId,
    }),
  },
  {
    method: 'POST',
    re: /\/page-member\/authorize$/,
    build: (_m, b) => ({
      action: 'AUTHORIZE_MEMBER',
      category: 'PAGE',
      label: 'Phân quyền thành viên trang',
      targetType: 'PAGE',
      targetId: () => b?.pageId,
    }),
  },
  {
    method: 'POST',
    re: /\/page\/post\/approve$/,
    build: (_m, b) => ({
      action: 'APPROVE_PAGE_POST',
      category: 'PAGE',
      label: 'Duyệt bài đăng trên trang',
      targetType: 'POST',
      targetId: () => b?.postId,
    }),
  },
  {
    method: 'POST',
    re: /\/page\/post\/remove$/,
    build: (_m, b) => ({
      action: 'REMOVE_PAGE_POST',
      category: 'PAGE',
      label: 'Gỡ bài đăng khỏi trang',
      targetType: 'POST',
      targetId: () => b?.postId,
    }),
  },
  // --- POST ---
  {
    method: 'DELETE',
    re: /\/posts\/([\w-]+)$/,
    build: (m) => ({
      action: 'DELETE_POST',
      category: 'POST',
      label: 'Xoá bài đăng',
      targetType: 'POST',
      targetId: () => m[1],
    }),
  },
  // --- STORY ---
  {
    method: 'DELETE',
    re: /\/admin\/stories\/([\w-]+)$/,
    build: (m) => ({
      action: 'DELETE_STORY',
      category: 'STORY',
      label: 'Xoá story',
      targetType: 'STORY',
      targetId: () => m[1],
    }),
  },
];

/** Các endpoint hệ thống không cần ghi log (nhiễu). */
const IGNORE_RE = [/\/auth\/refresh/, /\/auth\/me/, /upload/i, /presign/i, /generate-url/i];

export interface ActionInfo {
  action: string;
  category: AuditCategory;
  description: string;
  targetType?: string;
  targetId?: string | number;
}

/**
 * Suy ra hành động từ một request đã hoàn tất.
 * Trả về null nếu request không đáng ghi log.
 */
export function describeAction(
  method: string,
  url: string,
  body: any
): ActionInfo | null {
  const m = (method || 'GET').toUpperCase();
  if (m === 'GET' || m === 'HEAD' || m === 'OPTIONS') return null;
  if (IGNORE_RE.some((re) => re.test(url))) return null;

  let parsedBody: any = body;
  if (typeof body === 'string') {
    try {
      parsedBody = JSON.parse(body);
    } catch {
      parsedBody = body;
    }
  }

  for (const matcher of MATCHERS) {
    if (matcher.method !== m) continue;
    const match = url.match(matcher.re);
    if (!match) continue;
    const d = matcher.build(match, parsedBody);
    const targetId = d.targetId ? d.targetId(url, parsedBody) : undefined;
    return {
      action: d.action,
      category: d.category,
      description: d.label,
      targetType: d.targetType,
      targetId,
    };
  }

  // Hành động ghi (write) chưa được ánh xạ -> log chung để không bỏ sót
  const path = url.split('?')[0];
  return {
    action: `${m}_REQUEST`,
    category: 'SYSTEM',
    description: `${m} ${path}`,
  };
}

/* ------------------------------------------------------------------ */
/* Điểm vào cho interceptor                                            */
/* ------------------------------------------------------------------ */

export function recordRequestOutcome(opts: {
  method?: string;
  url?: string;
  body?: any;
  status: AuditStatus;
  statusCode?: number;
  errorMessage?: string;
}): void {
  if (!opts.url) return;
  const info = describeAction(opts.method || 'GET', opts.url, opts.body);
  if (!info) return;

  const actor = getCurrentActor();
  // Login thành công: lấy số điện thoại từ body làm actor
  let actorName = actor.name;
  let actorId = actor.id;
  if (info.action === 'LOGIN') {
    let b = opts.body;
    if (typeof b === 'string') {
      try {
        b = JSON.parse(b);
      } catch {
        /* ignore */
      }
    }
    if (b?.phone) actorName = b.phone;
  }

  addLog({
    actorType: 'ADMIN',
    actorId,
    actorName,
    action: info.action,
    description: info.description,
    category: info.category,
    targetType: info.targetType,
    targetId: info.targetId,
    method: (opts.method || 'GET').toUpperCase(),
    endpoint: opts.url.split('?')[0],
    status: opts.status,
    statusCode: opts.statusCode,
    meta: opts.errorMessage ? { error: opts.errorMessage } : undefined,
  });
}

/* ------------------------------------------------------------------ */
/* Truy vấn / lọc / xuất                                               */
/* ------------------------------------------------------------------ */

export interface AuditQuery {
  actorType?: AuditActorType | 'all';
  category?: AuditCategory | 'all';
  status?: AuditStatus | 'all';
  keyword?: string;
  from?: string; // yyyy-mm-dd
  to?: string; // yyyy-mm-dd
}

export function matchLog(log: AuditLog, q: AuditQuery): boolean {
  if (q.actorType && q.actorType !== 'all' && log.actorType !== q.actorType) return false;
  if (q.category && q.category !== 'all' && log.category !== q.category) return false;
  if (q.status && q.status !== 'all' && log.status !== q.status) return false;
  if (q.from && log.timestamp < new Date(q.from).toISOString()) return false;
  if (q.to) {
    const end = new Date(q.to);
    end.setHours(23, 59, 59, 999);
    if (log.timestamp > end.toISOString()) return false;
  }
  if (q.keyword) {
    const k = q.keyword.trim().toLowerCase();
    if (k) {
      const hay = [
        log.actorName,
        log.description,
        log.action,
        log.targetType,
        String(log.targetId ?? ''),
        log.targetName,
        log.endpoint,
      ]
        .filter(Boolean)
        .join(' ')
        .toLowerCase();
      if (!hay.includes(k)) return false;
    }
  }
  return true;
}

const auditLogService = {
  /** Lấy log cục bộ (đã lọc). */
  getLocal(query: AuditQuery = {}): AuditLog[] {
    return readAll().filter((l) => matchLog(l, query));
  },

  /**
   * Thử lấy log từ backend (`/admin/audit-logs`) — bao gồm cả hành động của
   * người dùng. Nếu backend chưa có endpoint, trả về log cục bộ.
   */
  async getLogs(query: AuditQuery = {}): Promise<{ logs: AuditLog[]; source: 'remote' | 'local' }> {
    try {
      const res = await axiosClient.get('/admin/audit-logs');
      const data = res?.data?.data ?? res?.data;
      if (Array.isArray(data) && data.length >= 0 && res.status < 400) {
        const logs = (data as AuditLog[]).filter((l) => matchLog(l, query));
        return { logs, source: 'remote' };
      }
    } catch {
      /* backend chưa hỗ trợ -> dùng cục bộ */
    }
    return { logs: this.getLocal(query), source: 'local' };
  },

  clear(): void {
    writeAll([]);
  },

  /** Xoá toàn bộ nhật ký trên máy chủ. */
  async clearRemote(): Promise<void> {
    await axiosClient.delete('/admin/audit-logs');
  },

  /** Xuất ra chuỗi CSV. */
  toCsv(logs: AuditLog[]): string {
    const headers = [
      'Thời gian',
      'Loại chủ thể',
      'Người thực hiện',
      'Hành động',
      'Mô tả',
      'Nhóm',
      'Đối tượng',
      'Mã đối tượng',
      'Phương thức',
      'Endpoint',
      'Trạng thái',
      'Mã HTTP',
    ];
    const escape = (v: any) => {
      const s = v == null ? '' : String(v);
      return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
    };
    const rows = logs.map((l) =>
      [
        new Date(l.timestamp).toLocaleString('vi-VN'),
        l.actorType,
        l.actorName,
        l.action,
        l.description,
        l.category,
        l.targetType ?? '',
        l.targetId ?? '',
        l.method ?? '',
        l.endpoint ?? '',
        l.status,
        l.statusCode ?? '',
      ]
        .map(escape)
        .join(',')
    );
    return '﻿' + [headers.join(','), ...rows].join('\n');
  },
};

export default auditLogService;

File: backend/src/main/java/iuh/fit/edu/backend/modules/audit/service/AuditLogService.java
/*
 * @ (#) AuditLogService.java    1.0
 * Copyright (c)  IUH. All rights reserved.
 */
package iuh.fit.edu.backend.modules.audit.service;

import iuh.fit.edu.backend.modules.audit.entity.AuditLog;
import iuh.fit.edu.backend.modules.audit.repository.AuditLogRepository;
import iuh.fit.edu.backend.modules.page.entity.Page;
import iuh.fit.edu.backend.modules.page.repository.PageRepository;
import iuh.fit.edu.backend.modules.user.entity.User;
import iuh.fit.edu.backend.modules.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/*
 * @description Ghi và truy vấn nhật ký hành động (audit log).
 * @author: Audit Log
 * @version: 1.0
 */
@Service
@RequiredArgsConstructor
public class AuditLogService {

    private final AuditLogRepository auditLogRepository;
    private final UserRepository userRepository;
    private final PageRepository pageRepository;
    private final AuditLogStream auditLogStream;

    /**
     * Chỉ bỏ qua chính endpoint audit-log (đọc + stream) để tránh vòng lặp tự
     * khuếch đại. Mọi API khác — kể cả GET — đều được ghi log.
     */
    private static final Pattern[] IGNORE = {
            Pattern.compile("/admin/audit-logs"),
            // Luồng SSE báo cáo giữ kết nối lâu, không cần ghi log
            Pattern.compile("/admin/reports/stream"),
    };

    /* ----- Kết quả ánh xạ hành động ----- */
    public record ActionInfo(String action, String category, String description,
                             String targetType, String targetId) {}

    private interface Builder {
        ActionInfo build(Matcher m);
    }

    private record Rule(String method, Pattern pattern, Builder builder) {}

    /** Bảng ánh xạ method + path -> hành động có ý nghĩa (đồng bộ với frontend-admin). */
    private static final List<Rule> RULES = List.of(
            // AUTH
            rule("POST", "/auth/login$", m -> new ActionInfo("LOGIN", "AUTH", "Đăng nhập", null, null)),
            rule("POST", "/auth/logout$", m -> new ActionInfo("LOGOUT", "AUTH", "Đăng xuất", null, null)),
            rule("POST", "/auth/register$", m -> new ActionInfo("REGISTER", "AUTH", "Đăng ký tài khoản", null, null)),
            rule("POST", "/auth/reset-password$", m -> new ActionInfo("RESET_PASSWORD", "AUTH", "Đặt lại mật khẩu", null, null)),
            // USER / ADMIN
            rule("POST", "/admin/lock/(\\w+)", m -> new ActionInfo("LOCK_USER", "USER", "Khoá tài khoản người dùng", "USER", m.group(1))),
            rule("POST", "/admin/unlock/(\\w+)", m -> new ActionInfo("UNLOCK_USER", "USER", "Mở khoá tài khoản người dùng", "USER", m.group(1))),
            rule("DELETE", "/auth/users/(\\w+)", m -> new ActionInfo("DELETE_USER", "USER", "Xoá tài khoản người dùng", "USER", m.group(1))),
            rule("PUT", "/auth/users/(\\w+)", m -> new ActionInfo("UPDATE_USER", "USER", "Cập nhật thông tin người dùng", "USER", m.group(1))),
            // FRIEND
            rule("POST", "/friends/request", m -> new ActionInfo("FRIEND_REQUEST", "USER", "Gửi lời mời kết bạn", "USER", null)),
            rule("POST", "/friends/accept", m -> new ActionInfo("FRIEND_ACCEPT", "USER", "Chấp nhận kết bạn", "USER", null)),
            rule("POST", "/friends/block", m -> new ActionInfo("BLOCK_USER", "USER", "Chặn người dùng", "USER", null)),
            // PAGE
            rule("DELETE", "/page/delete/(\\w+)", m -> new ActionInfo("DELETE_PAGE", "PAGE", "Xoá trang", "PAGE", m.group(1))),
            rule("POST", "/page/create", m -> new ActionInfo("CREATE_PAGE", "PAGE", "Tạo trang", "PAGE", null)),
            rule("POST", "/page-member/approve-join$", m -> new ActionInfo("APPROVE_MEMBER", "PAGE", "Duyệt yêu cầu tham gia trang", "PAGE", null)),
            rule("POST", "/page-member/reject-join$", m -> new ActionInfo("REJECT_MEMBER", "PAGE", "Từ chối yêu cầu tham gia trang", "PAGE", null)),
            rule("POST", "/page-member/block$", m -> new ActionInfo("BLOCK_MEMBER", "PAGE", "Chặn thành viên trang", "PAGE", null)),
            rule("POST", "/page-member/cancel-block$", m -> new ActionInfo("UNBLOCK_MEMBER", "PAGE", "Bỏ chặn thành viên trang", "PAGE", null)),
            rule("POST", "/page-member/delete$", m -> new ActionInfo("REMOVE_MEMBER", "PAGE", "Xoá thành viên khỏi trang", "PAGE", null)),
            rule("POST", "/page-member/authorize$", m -> new ActionInfo("AUTHORIZE_MEMBER", "PAGE", "Phân quyền thành viên trang", "PAGE", null)),
            rule("POST", "/page/post/approve$", m -> new ActionInfo("APPROVE_PAGE_POST", "PAGE", "Duyệt bài đăng trên trang", "POST", null)),
            rule("POST", "/page/post/remove$", m -> new ActionInfo("REMOVE_PAGE_POST", "PAGE", "Gỡ bài đăng khỏi trang", "POST", null)),
            // POST
            rule("POST", "/posts$", m -> new ActionInfo("CREATE_POST", "POST", "Tạo bài đăng", "POST", null)),
            rule("DELETE", "/posts/([\\w-]+)$", m -> new ActionInfo("DELETE_POST", "POST", "Xoá bài đăng", "POST", m.group(1))),
            rule("POST", "/reactions", m -> new ActionInfo("REACT_POST", "POST", "Bày tỏ cảm xúc", "POST", null)),
            rule("POST", "/comments", m -> new ActionInfo("CREATE_COMMENT", "POST", "Bình luận", "POST", null)),
            // STORY
            rule("POST", "/stories$", m -> new ActionInfo("CREATE_STORY", "STORY", "Đăng story", "STORY", null)),
            rule("DELETE", "/admin/stories/([\\w-]+)$", m -> new ActionInfo("DELETE_STORY", "STORY", "Xoá story", "STORY", m.group(1))),
            // REPORT
            rule("POST", "/report$", m -> new ActionInfo("SUBMIT_REPORT", "REPORT", "Gửi báo cáo", null, null)),
            rule("POST", "/admin/reports/(\\d+)/handle$", m -> new ActionInfo("HANDLE_REPORT", "REPORT", "Xử lý báo cáo", "REPORT", m.group(1)))
    );

    private static Rule rule(String method, String regex, Builder b) {
        return new Rule(method, Pattern.compile(regex), b);
    }

    /** Suy ra hành động từ request; trả về null nếu không đáng ghi log. */
    public ActionInfo describe(String method, String path) {
        if (method == null) return null;
        String m = method.toUpperCase();
        // OPTIONS chỉ là preflight CORS, không phải hành động thật
        if (m.equals("OPTIONS")) return null;
        for (Pattern ig : IGNORE) {
            if (ig.matcher(path).find()) return null;
        }
        for (Rule r : RULES) {
            if (!r.method().equals(m)) continue;
            Matcher matcher = r.pattern().matcher(path);
            if (matcher.find()) {
                return r.builder().build(matcher);
            }
        }
        // Request chưa được ánh xạ (bao gồm các lượt đọc GET) -> log chung để không bỏ sót.
        // Lấy "đối tượng" từ path: phân đoạn tài nguyên đầu tiên + id (nếu có) sau /api/.
        String resource = null;
        String tid = null;
        String stripped = path.replaceFirst("^/api/", "");
        String[] segs = stripped.split("/");
        if (segs.length > 0 && !segs[0].isBlank()) resource = segs[0];
        if (segs.length > 1 && segs[1].matches("\\d+")) tid = segs[1];

        // Mô tả = phương thức HTTP + path (GET/POST/PUT/DELETE/PATCH ...)
        String desc = m + " " + path;
        return new ActionInfo(m + "_REQUEST", "SYSTEM", desc, resource, tid);
    }

    /**
     * Ghi một bản ghi nhật ký. Chạy bất đồng bộ để không ảnh hưởng request gốc.
     */
    @Async
    @Transactional
    public void record(String method, String path, int statusCode,
                       String actorPhone, boolean fromAdminConsole, String errorMessage) {
        try {
            ActionInfo info = describe(method, path);
            if (info == null) return;

            String actorType;
            Long actorId = null;
            String actorName;

            User actor = (actorPhone != null) ? userRepository.findByPhone(actorPhone) : null;
            if (actor != null) {
                actorId = actor.getId();
                actorName = (actor.getName() != null && !actor.getName().isBlank())
                        ? actor.getName()
                        : (actor.getUsername() != null ? actor.getUsername() : actor.getPhone());
            } else {
                actorName = actorPhone != null ? actorPhone : "(ẩn danh)";
            }

            if (fromAdminConsole) {
                actorType = "ADMIN";
            } else if (actorPhone != null) {
                actorType = "USER";
            } else {
                actorType = "SYSTEM";
            }

            // Tra cứu tên đối tượng (nếu là USER/PAGE và có id dạng số)
            String targetName = resolveTargetName(info.targetType(), info.targetId());

            AuditLog log = AuditLog.builder()
                    .timestamp(OffsetDateTime.now())
                    .actorType(actorType)
                    .actorId(actorId)
                    .actorName(actorName)
                    .action(info.action())
                    .description(info.description())
                    .category(info.category())
                    .targetType(info.targetType())
                    .targetId(info.targetId())
                    .targetName(targetName)
                    .method(method.toUpperCase())
                    .endpoint(path)
                    .status(statusCode >= 400 ? "FAILED" : "SUCCESS")
                    .statusCode(statusCode)
                    .meta(errorMessage != null ? "{\"error\":\"" + errorMessage.replace("\"", "'") + "\"}" : null)
                    .build();

            auditLogRepository.save(log);

            // Đẩy realtime tới các trang quản trị đang mở
            auditLogStream.broadcast(log);
        } catch (Exception ignored) {
            // Ghi log không bao giờ được làm hỏng request gốc
        }
    }

    /** Tra cứu tên hiển thị của đối tượng bị tác động (USER/PAGE). */
    private String resolveTargetName(String targetType, String targetId) {
        if (targetType == null || targetId == null) return null;
        try {
            Long id = Long.valueOf(targetId);
            switch (targetType.toUpperCase()) {
                case "USER" -> {
                    User u = userRepository.findById(id).orElse(null);
                    if (u == null) return null;
                    return (u.getName() != null && !u.getName().isBlank()) ? u.getName() : u.getUsername();
                }
                case "PAGE" -> {
                    Page p = pageRepository.findById(id).orElse(null);
                    return p != null ? p.getName() : null;
                }
                default -> {
                    return null;
                }
            }
        } catch (NumberFormatException ex) {
            return null;
        }
    }

    /** Xoá toàn bộ nhật ký. */
    @Transactional
    public void clearAll() {
        auditLogRepository.deleteAllInBatch();
        auditLogStream.broadcastCleared();
    }

    /** Lấy các bản ghi mới nhất. */
    @Transactional(readOnly = true)
    public List<AuditLog> getRecent(int limit) {
        int size = Math.min(Math.max(limit, 1), 2000);
        return auditLogRepository.findAllByOrderByTimestampDesc(PageRequest.of(0, size));
    }
}
